tests/test_all_gather.py

```diff
@@ -0,0 +1,38 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+
+import torch
+import poptorch
+import poptorch_experimental_addons as pea
+import random
+import numpy as np
+import einops
+
+
+def set_seed(seed: int):
+    torch.manual_seed(seed)
+    random.seed(seed)
+    np.random.seed(seed)
+
+
+class _ReplicatedAllGather(torch.nn.Module):
+    def __init__(self, replication_factor: int = 1):
+        super().__init__()
+        self.replication_factor = replication_factor
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return pea.collectives.replicated_all_gather(x, self.replication_factor)
+
+
+def test_all_gather():
+    set_seed(112358)
+    x = torch.randn(32, 24)
```

Comment:
The idea is that `x` is initially sharded across IPUs? Perhaps clearer to do `torch.randn(2 * dp, 24)`?

While here, maybe I'd suggest `dp -> n_ipu` or something?